### PR TITLE
Drug stock api

### PR DESCRIPTION
**Pull request by harimohanraj89**
**Story card:** [sc-7687](https://app.shortcut.com/simpledotorg/story/7687/drug-stock-api)

This PR introduces a `/api/v4/drug_stocks` API endpoint for the mobile app to query in order to power the drug stock reminder nudges in the app.

* Accepts requests for a specific month, and uses the current user's facility (X-Facility-ID header)
* Returns a 200 with drug stock info if a drug stock report has been submitted for that facility and month
* Returns a 400 of no report has been submitted yet

### To Do

* [x] Fix the API specs and update documentation

### CODE REVIEW

File: app/controllers/api/v4/drug_stocks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    @drug_stocks = DrugStock.latest_for_facilities([current_facility], for_end_of_month)
+
+    head :not_found if @drug_stocks.empty?
+  end
+
+  private
+
+  def for_end_of_month
+    @for_end_of_month ||= Date.parse(params[:date]).end_of_month
```
Comment (prabhanshuguptagit) [inline app/controllers/api/v4/drug_stocks_controller.rb:16]: `Date.parse` accepts a wide range of date formats as valid input. Let's use `Date.strptime` and specify the exact format.

File: app/controllers/api/v4/drug_stocks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  rescue Date::Error, TypeError
```
Comment (prabhanshuguptagit) [inline app/controllers/api/v4/drug_stocks_controller.rb:17]: Is the `TypeError` to rescue non-string params being passed? Should we coerce `params[:date]` to a string first in that case?

File: app/controllers/api/v4/drug_stocks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+class Api::V4::DrugStocksController < APIController
+  def index
+    if for_end_of_month.nil?
+      render status: 400, json: {errors: ["valid `date` must be provided - eg. 2021-10-29"]}
```
Comment (prabhanshuguptagit) [inline app/controllers/api/v4/drug_stocks_controller.rb:4]: Shall we also specify the date format we expect in this (already very helpful) error message?

File: spec/controllers/api/v4/drug_stocks_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      it "returns a bad request if date is missing" do
+        request.headers["X-User-Id"] = user.id
+        request.headers["X-Facility-Id"] = facility.id
+        request.headers["Authorization"] = "Bearer #{user.access_token}"
+
+        get :index
+
+        expect(response.status).to eq(400)
+        expect(response.body).to include("date")
+      end
```
Comment (prabhanshuguptagit) [inline spec/controllers/api/v4/drug_stocks_controller_spec.rb:63]: 👌🏼

File: spec/controllers/api/v4/drug_stocks_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        "drugs" => drug_stocks.map do |stock|
+          {
+            "protocol_drug_id" => stock.protocol_drug_id,
+            "in_stock" => stock.in_stock,
+            "received" => stock.received
+          }
+        end
+      }
+
+      allow(DrugStock).to receive(:latest_for_facilities).with([facility], end_of_month).and_return(drug_stocks)
```
Comment (prabhanshuguptagit) [inline spec/controllers/api/v4/drug_stocks_controller_spec.rb:28]: Is this mocking needed if the drug stocks have been created in the right month at `facility`?

**Review by prabhanshuguptagit — APPROVED** [review]
Nifty :shipit:

File: app/controllers/api/v4/drug_stocks_controller.rb (hunk 2 above)
Comment (harimohanraj89) [reply app/controllers/api/v4/drug_stocks_controller.rb:17]: The TypeError is to catch `nil` - ie, date was not provided.

File: spec/controllers/api/v4/drug_stocks_controller_spec.rb (hunk 5 above)
Comment (harimohanraj89) [reply spec/controllers/api/v4/drug_stocks_controller_spec.rb:28]: Ah it is indeed not. I will remove.

**Comment by harimohanraj89** [comment]
Thank you for the review, addressed feedback in [9f892ab](https://github.com/simpledotorg/simple-server/pull/3635/commits/9f892ab2f74fc47774923863d41084ddc24b1571) 👍

**Pull request merged**